pubsub/awssnssqs/awssnssqs.go

```diff
@@ -448,45 +582,74 @@ func openSubscription(ctx context.Context, sess client.ConfigProvider, qURL stri
 // ReceiveBatch implements driver.Subscription.ReceiveBatch.
 func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
 	output, err := s.client.ReceiveMessageWithContext(ctx, &sqs.ReceiveMessageInput{
-		QueueUrl:            aws.String(s.qURL),
-		MaxNumberOfMessages: aws.Int64(int64(maxMessages)),
+		QueueUrl:              aws.String(s.qURL),
+		MaxNumberOfMessages:   aws.Int64(int64(maxMessages)),
+		MessageAttributeNames: []*string{aws.String("All")},
 	})
 	if err != nil {
 		return nil, err
 	}
 	var ms []*driver.Message
 	for _, m := range output.Messages {
-		type MsgBody struct {
-			Message           string
-			MessageAttributes map[string]struct{ Value string }
-		}
-		var body MsgBody
-		if err := json.Unmarshal([]byte(*m.Body), &body); err != nil {
-			return nil, err
+		// By default, messages from SNS come with everything in a JSON-encoded Body.
+		// For example, message attributes are not in m.MessageAttributes, but
+		// are rather in a MessageAttributes field in the JSON-encoded body.
+		// If RawMessageDelivery is enabled as part of the SQS queue subscription
+		// to the SNS topic, then body is in Body and attributes are in
+		// MessageAttributes. Same deal if you send directly to the SQS queue.
+		//
+		// If it looks like the body might be JSON, we try decoding it. If it
+		// doesn't look like JSON, or if the JSON decode fails, we use the raw
+		// data from m.
+		bodyStr := aws.StringValue(m.Body)
+		rawAttrs := map[string]string{}
+		useJSON := len(m.MessageAttributes) == 0 // if we got attributes, it's raw
+		if useJSON {
+			type MsgBody struct {
+				Message           string
+				MessageAttributes map[string]struct{ Value string }
+			}
+			var bodyJSON MsgBody
+			if err := json.Unmarshal([]byte(bodyStr), &bodyJSON); err == nil {
+				// JSON decode succeeded; get attributes from the decoded struct,
+				// and update the body to be the JSON Message field.
+				for k, v := range bodyJSON.MessageAttributes {
+					rawAttrs[k] = v.Value
+				}
+				bodyStr = bodyJSON.Message
+			} else {
+				// JSON decode failed; leave bodyStr alone. There can't be any
+				// attributes.
+			}
+		} else {
+			for k, v := range m.MessageAttributes {
+				rawAttrs[k] = aws.StringValue(v.StringValue)
+			}
 		}
-		// See BodyBase64Encoding for details on when we base64 decode message bodies.
+
 		decodeIt := false
 		attrs := map[string]string{}
-		for k, v := range body.MessageAttributes {
+		for k, v := range rawAttrs {
```

Comment:
(Re the line above.) The doc comment for `Message.Metadata` says it will be nil if there are no attributes, but in fact it will be an empty map.

Reply:
The portable type sets it to `nil` if empty:
https://github.com/google/go-cloud/blob/master/pubsub/pubsub.go#L606-L608